### ISSUE TITLE
Add Builder on some classes to simplify some unit tests

### ISSUE THREAD

**Issue by alexandrenavarro** [spring-projects/spring-statemachine#526]
Add Builder on some classes to simplify some unit tests.

When you want to unit test Action / Guard, the creation of StateContext / DefaultStateContext is quite painful because of the constructor with a lot of arguments (see Java Effective Book about it).

To simplify it, the best solution in java is to add Builder.
There are 2 libs to create quickly builders in java which can generate Builder at compile time
1) lombok 
2) pojobuilder

1) lombok is used a lot in java community for generating getter/setter/builder/toString/ ... . 
The main drawback is the generation is done in the same file, so you need to install plugin to build in a IDE. As I don't see it in the project, I supposed you may not want to add it in your project?

2) pojobuilder is used only to generate builder in a different class. 
It does only one thing but it does it well.

If you are ok with the enhancement to add some builders notably on DefaultStateContext but maybe some other classes if it is needed, tell me which lib you prefer to use, I will propose you a pull request.

**Comment by jvalkeal** [comment]
Yes, sound reasonable as I've also felt a same pain. I know that some people love lombok and other similar stuff but I'm absolutely gutted for code generation tools like those. I prefer doing things manually as then you have full control and you don't need any IDE support(which usually are pretty bad and error prone).

**Comment by alexandrenavarro** [comment]
Hi,

I did a pull request https://github.com/spring-projects/spring-statemachine/pull/535 with pojobuilder to create automatically builders on some classes via annotation.
Nevertheless I'm not totally satisfied with what I did.
The builder is correctly generated, the class is perfectly generated and useable when you use the lib but the generated .java is generated in the same directory class in build/class and not in seperated directory. I know very well maven (I know how to do it in maven) nevertheless I don't know a lot of things in gradle, I think it is totally possible but I don't know yet how.

What I proposed you is 3 potential solutions :
1. just take the source code generated and add in your source code.
2. The pull request is ok for you, merge as it if you are ok.
3. I can dig in gradle to understand how to do it with a generated java folder and works correctly with IDE directly inside the spring-statemachine projects.

Tell me if you are ok with one of the solution, if it is 3, any help if you know well gradle is welcomed.
Personally, I prefer solution 3 but I don't know yet how to do it correcly in gradle yet.
